test/michael_scott_queue/stm_michael_scott_queue.ml

```diff
@@ -1,65 +1,73 @@
+    type sut = int Ms_queue.t
 
-  type state = int list
-  type sut = int Ms_queue.t
+    let arb_cmd _s =
+      let int_gen = Gen.nat in
+      QCheck.make ~print:show_cmd
+        (Gen.oneof
+           [
+             Gen.map (fun i -> Push i) int_gen;
+             Gen.return Pop;
+             Gen.return Peek;
+             Gen.return Is_empty;
+           ])
 
-  let arb_cmd _s =
-    let int_gen = Gen.nat in
-    QCheck.make ~print:show_cmd
-      (Gen.oneof
-         [
-           Gen.map (fun i -> Push i) int_gen;
-           Gen.return Pop;
-           Gen.return Peek;
-           Gen.return Is_empty;
-         ])
+    let init_state = []
+    let init_sut () = Ms_queue.create ()
+    let cleanup _ = ()
 
-  let init_state = []
-  let init_sut () = Ms_queue.create ()
-  let cleanup _ = ()
+    let next_state c s =
+      match c with
+      | Push i -> i :: s
+      | Pop -> ( match List.rev s with [] -> s | _ :: s' -> List.rev s')
+      | Peek | Is_empty -> s
 
-  let next_state c s =
-    match c with
-    | Push i -> i :: s
-    | Pop -> ( match List.rev s with [] -> s | _ :: s' -> List.rev s')
-    | Peek | Is_empty -> s
+    let precond _ _ = true
 
-  let precond _ _ = true
+    let run c d =
+      match c with
+      | Push i -> Res (unit, Ms_queue.push d i)
+      | Pop -> Res (result int exn, protect Ms_queue.pop_exn d)
+      | Peek -> Res (result int exn, protect Ms_queue.peek_exn d)
+      | Is_empty -> Res (bool, Ms_queue.is_empty d)
 
-  let run c d =
-    match c with
-    | Push i -> Res (unit, Ms_queue.push d i)
-    | Pop -> Res (result int exn, protect Ms_queue.pop d)
-    | Peek -> Res (result int exn, protect Ms_queue.peek d)
-    | Is_empty -> Res (bool, Ms_queue.is_empty d)
+    let postcond c (s : state) res =
+      match (c, res) with
+      | Push _, Res ((Unit, _), _) -> true
+      | (Pop | Peek), Res ((Result (Int, Exn), _), res) -> (
+          match List.rev s with
+          | [] -> res = Error Ms_queue.Empty
+          | j :: _ -> res = Ok j)
+      | Is_empty, Res ((Bool, _), res) -> res = (s = [])
+      | _, _ -> false
+  end
 
-  let postcond c (s : state) res =
-    match (c, res) with
-    | Push _, Res ((Unit, _), _) -> true
-    | (Pop | Peek), Res ((Result (Int, Exn), _), res) -> (
-        match List.rev s with
-        | [] -> res = Error Ms_queue.Empty
-        | j :: _ -> res = Ok j)
-    | Is_empty, Res ((Bool, _), res) -> res = (s = [])
-    | _, _ -> false
+  let run () =
+    Stm_run.run ~count:500 ~verbose:true
+      ~name:("Saturn_lockfree." ^ Queue.name)
+      (module Spec)
 end
 
 let () =
-  Stm_run.run ~count:500 ~verbose:true
-    ~name:"Saturn_lockfree.Michael_scott_queue"
-    (module Spec)
-  |> exit
+  let module Safe = STM_ms_queue (Ms_queues.Michael_scott_queue) in
+  let exit_code = Safe.run () in
+  if exit_code <> 0 then exit exit_code
+  else
+    let module Unsafe = STM_ms_queue (Ms_queues.Michael_scott_queue_unsafe) in
+    Unsafe.run () |> exit
```

Comment:
This could also be a bit less verbose using first-class modules.  We are not really interested in the `Safe` and `Unsafe` modules as such.  We just want to run the tests and get an int as a result.